types/styled-components/index.d.ts

```diff
@@ -132,101 +254,135 @@ export interface BaseThemedCssFunction<T> {
     ): Keyframes;
 
+    createGlobalStyle(cssObject: CSSObject): GlobalStyleComponent<{}, T>;
     createGlobalStyle<P = {}>(
         strings: TemplateStringsArray,
         ...interpolations: Array<Interpolation<ThemedStyledProps<P, T>>>
-    ): GlobalStyleClass<P, T>;
+    ): GlobalStyleComponent<P, T>;
+    createGlobalStyle<P = {}>(
+        func: InterpolationFunction<ThemedStyledProps<P, T>>
+    ): GlobalStyleComponent<P, T>;
 
     withTheme: WithThemeFnInterface<T>;
-
     ThemeProvider: ThemeProviderComponent<T>;
-    ThemeConsumer: ThemeConsumerComponent<T>;
+    ThemeConsumer: React.Consumer<T>;
+    ThemeContext: React.Context<T>;
 }
 
 declare const styled: StyledInterface;
 
 export const css: ThemedCssFunction<DefaultTheme>;
 
-export type BaseWithThemeFnInterface<T> = <P extends { theme?: T }>(
-        component: React.ComponentType<P>,
-    ) => React.ComponentClass<WithOptionalTheme<P, T>>;
-export type WithThemeFnInterface<T> = BaseWithThemeFnInterface<Extract<keyof T, string> extends never ? any : T>;
+export type BaseWithThemeFnInterface<T extends object> = <
+    C extends React.ComponentType<any>
+>(
+    // this check is roundabout because the extends clause above would
+    // not allow any component that accepts _more_ than theme as a prop
+    component: React.ComponentProps<C> extends { theme?: T } ? C : never
+) => React.ForwardRefExoticComponent<
+    WithOptionalTheme<React.ComponentPropsWithRef<C>, T>
+>;
+export type WithThemeFnInterface<T extends object> = BaseWithThemeFnInterface<
+    Extract<keyof T, string> extends never ? any : T
+>;
 export const withTheme: WithThemeFnInterface<DefaultTheme>;
 
-export interface ThemeConsumerProps<T> {
-    children(theme: T): React.ReactNode;
+// tslint:disable-next-line:no-empty-interface
+export interface DefaultTheme {}
+
+export interface ThemeProviderProps<T extends object> {
+    children?: React.ReactChild; // only one child is allowed, goes through React.Children.only
+    theme: T | ((theme: T) => T);
 }
-export type BaseThemeConsumerComponent<T> = React.ComponentClass<ThemeConsumerProps<T>>;
-export type ThemeConsumerComponent<T> = BaseThemeConsumerComponent<Extract<keyof T, string> extends never ? any : T>;
-export const ThemeConsumer: ThemeConsumerComponent<object>;
+export type BaseThemeProviderComponent<T extends object> = React.ComponentClass<
+    ThemeProviderProps<T>
+>;
+export type ThemeProviderComponent<
+    T extends object
+> = BaseThemeProviderComponent<Extract<keyof T, string> extends never ? any : T>;
+export const ThemeProvider: ThemeProviderComponent<DefaultTheme>;
+// NOTE: this technically starts as undefined
+// Also, this cannot be DefaultTheme as it breaks TypedComponents' assignability
+export const ThemeContext: React.Context<any>;
+export const ThemeConsumer: typeof ThemeContext['Consumer'];
 
 export interface Keyframes {
     getName(): string;
 }
 
+export function keyframes(cssObject: CSSObject): Keyframes;
```

Comment:
I don't know if `keyframes` accepts the `InterpolationFunction` variation. It does have `SimpleInterpolation[]` in the second overload.

(actually, I haven't tested this either; more homework...)